src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,545 @@
+	}
+	return true;
+}
+
+
+/*
+ * HasConstantFilterOnUniqueColumn returns true if the given rangeTableEntry has a constant
+ * filter on a unique column.
+ */
+static bool
+HasConstantFilterOnUniqueColumn(RangeTblEntry *rangeTableEntry,
+								RelationRestriction *relationRestriction)
+{
+	if (rangeTableEntry == NULL)
+	{
+		return false;
+	}
+	List *baseRestrictionList = relationRestriction->relOptInfo->baserestrictinfo;
+	List *restrictClauseList = get_all_actual_clauses(baseRestrictionList);
+	if (ContainsFalseClause(restrictClauseList))
+	{
+		/* If there is a WHERE FALSE, we consider it as a constant filter. */
+		return true;
+	}
+	List *rteEqualityColumnsNos =
+		FetchEqualityAttrNumsForRTE((Node *) restrictClauseList);
+
+	List *uniqueIndexColumnsList = ExecuteFunctionOnEachTableIndex(rangeTableEntry->relid,
+																   AppendUniqueIndexColumnsToList);
+	IndexColumns *indexColumns = NULL;
+	foreach_ptr(indexColumns, uniqueIndexColumnsList)
+	{
+		List *uniqueIndexColumnNos = indexColumns->indexColumnNos;
+		if (FirstIsSuperSetOfSecond(rteEqualityColumnsNos,
+									uniqueIndexColumnNos))
+		{
+			return true;
+		}
+	}
+	return false;
+}
+
+
+/*
+ * FirstIsSuperSetOfSecond returns true if the first int List
+ * contains every element of the second int List.
+ */
+static bool
+FirstIsSuperSetOfSecond(List *firstIntList, List *secondIntList)
+{
+	int curInt = 0;
+	foreach_int(curInt, secondIntList)
+	{
+		if (!list_member_int(firstIntList, curInt))
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+/*
+ * AppendUniqueIndexColumnsToList adds the given index's column numbers if it is a
+ * unique index.
+ */
+static void
+AppendUniqueIndexColumnsToList(Form_pg_index indexForm, List **uniqueIndexGroups)
+{
+	if (indexForm->indisunique || indexForm->indisprimary)
+	{
+		IndexColumns *indexColumns = palloc0(sizeof(IndexColumns));
+		List *uniqueIndexes = NIL;
+		for (int i = 0; i < indexForm->indkey.dim1; i++)
+		{
+			uniqueIndexes = list_append_unique_int(uniqueIndexes,
+												   indexForm->indkey.values[i]);
+		}
+		if (list_length(uniqueIndexes) == 0)
+		{
+			return;
+		}
+		indexColumns->indexColumnNos = uniqueIndexes;
+		*uniqueIndexGroups = lappend(*uniqueIndexGroups, indexColumns);
+	}
+}
+
+
+/*
+ * RequiredAttrNumbersForRelation returns the required attribute numbers for
+ * the input RTE relation in order for the planning to succeed.
+ *
+ * The function could be optimized by not adding the columns that only appear
+ * WHERE clause as a filter (e.g., not a join clause).
+ */
+static List *
+RequiredAttrNumbersForRelation(RangeTblEntry *rangeTableEntry,
+							   PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RelationRestriction *relationRestriction =
+		RelationRestrictionForRelation(rangeTableEntry, plannerRestrictionContext);
+
+	if (relationRestriction == NULL)
+	{
+		return NIL;
+	}
+
+	PlannerInfo *plannerInfo = relationRestriction->plannerInfo;
+	Query *queryToProcess = plannerInfo->parse;
```

Comment:
There's something slightly odd about taking the top-level parse tree here, e.g. what if we're looking at a subquery? Can we never look at a subquery?

Reply:
For now, I'm ok with adding a comment here on why we use it.